protocol/icmpv6.go

```diff
@@ -0,0 +1,628 @@
+//    *                                                               *
+//    |                                                               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |                                                               |
+//    .                                                               .
+//    .                         Auxiliary Data                        .
+//    .                                                               .
+//    |                                                               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+type MLDv2Record struct {
+	Type             uint8
+	AuxDataLen       uint8 // this may be 0 to indicate the absence of any auxiliary data.
+	NumberOfSources  uint16
+	MulticastAddress net.IP
+	SourceAddresses  []net.IP
+	AuxData          []uint32 // NOT USED
+}
+
+func (r *MLDv2Record) Len() uint16 {
+	return 20 + uint16(r.AuxDataLen)*4 + r.NumberOfSources*16
+}
+
+func (r *MLDv2Record) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, int(r.Len()))
+	n := 0
+	data[n] = r.Type
+	n += 1
+	data[n] = r.AuxDataLen
+	n += 1
+	binary.BigEndian.PutUint16(data[n:], r.NumberOfSources)
+	n += 2
+	copy(data[n:n+4], r.MulticastAddress.To16())
+	n += 16
+	for _, src := range r.SourceAddresses {
+		copy(data[n:n+4], src.To16())
+		n += 16
+	}
+	for _, d := range r.AuxData {
+		binary.BigEndian.PutUint32(data[n:], d)
+		n += 4
+	}
+	return
+}
+
+func (r *MLDv2Record) UnmarshalBinary(data []byte) error {
+	if len(data) < 8 {
+		return errors.New("The []byte is too short to unmarshal a full MLDv2Record message.")
+	}
+	n := 0
+	r.Type = data[n]
+	n += 1
+	r.AuxDataLen = data[n]
+	n += 1
+	r.NumberOfSources = binary.BigEndian.Uint16(data[n:])
+	n += 2
+	r.MulticastAddress = data[n : n+16]
+	n += 16
+	if len(data) < int(r.Len()) {
+		return fmt.Errorf("The []byte is too short to unmarshal a full MLDv2Record message.")
+	}
+	for i := uint16(0); i < r.NumberOfSources; i++ {
+		r.SourceAddresses = append(r.SourceAddresses, data[n:n+16])
+		n += 16
+	}
+	for i := uint8(0); i < r.AuxDataLen; i++ {
+		r.AuxData = append(r.AuxData, binary.BigEndian.Uint32(data[n:]))
+		n += 4
+	}
+	return nil
+}
+
+func NewMLDv2Record(recordType uint8, group net.IP, sources []net.IP) *MLDv2Record {
+	return &MLDv2Record{
+		Type:             recordType,
+		MulticastAddress: group,
+		NumberOfSources:  uint16(len(sources)),
+		SourceAddresses:  sources,
+	}
+}
+
+func NewICMPv6ByHeader(header *ICMPv6Header) util.Message {
+	switch header.Type {
+	case ICMPv6_Type_EchoRequest:
+		fallthrough
+	case ICMPv6_Type_EchoReply:
+		return new(ICMPv6EchoReqRpl)
+	case ICMPv6_Type_MLD_Query:
+		fallthrough
+	case ICMPv6_Type_MLD_Report:
+		fallthrough
+	case ICMPv6_Type_MLD_Done:
+		return new(MLD)
+	case ICMPv6_Type_MLDv2_Report:
+		return new(MLDv2Report)
+	}
+	return new(util.Buffer)
```

Comment:
can you add a comment for this "fallthrough" case?

Reply:
Added, and it is actually because MLD Query, Report and Done are using the same struct. 

But I actually have a question on differentiating MLD Query and MLDv2 Query, they are actually using the same Type, but having different structs. Could you help give some suggestions on this?

Reply:
> But I actually have a question on differentiating MLD Query and MLDv2 Query, they are actually using the same Type, but having different structs. Could you help give some suggestions on this?

I guess you can use the length of the data to differentiate between the 2? You could add the length as a function parameter. Alternatively you can use the same struct and initialize the extra v2 fields when you unmarshal based on whether there is data remaining. The single struct can have a method for users to distinguish between v1 and v2.

Reply:
Is this resolved? I don't see `NewMLDv2Query` called anywhere

Reply:
Not resolved yet. I am thinking shall I use a struct dedicated for MDLQuery, which has a single bool field to mark if this is for MLDv2. If yes, we could marshal the MLDv2 speicific fields in the bytes, and check data length to decide the MLD version when unmarshal the bytes. What do you think? @antoninbas

Reply:
Sounds good. I think it matches my second idea above?

> Alternatively you can use the same struct and initialize the extra v2 fields when you unmarshal based on whether there is data remaining. The single struct can have a method for users to distinguish between v1 and v2.

Reply:
Yes.